### ISSUE TITLE
PDFs not displaying

### ISSUE THREAD

**Issue by oraclesod** [mkucej/i-librarian-free#42]
I have recently inherited an old i-librarian server that was running v2, i have since then migrated it to a ubuntu 20.04 box and upgraded it to 4.10 and its happy. When i try to upgrade to 5.5 however i run into issues.

The database appears to migrate successfully and I ran the PDF extraction, which seems to also work (other than lots of Syntax Warning in the apache error log). I can view the various items in the library, however the PDF and Download buttons are greyed out. If i go into an item and select PDF->Open it fails horribly, i get the PDF app window, but then lots of 500 errors start appearing in the browser console log (see attached pic)

<img width="1789" alt="Screen Shot 2020-10-03 at 11 11 21 PM" src="https://user-images.githubusercontent.com/1288780/95008449-cf0bf000-05ce-11eb-9c0a-fdb5721347d4.png">

Not sure what to do at this point or where this issue is coming from.

**Comment by mkucej** [comment]
I wonder if you have `file_hash` values populated in the `items`  SQLite table?

**Comment by oraclesod** [comment]
Just checked, they are all Null.....

**Comment by mkucej** [comment]
OK. The `file_hash` (md5 hashes of PDFs) values were not created during your upgrade from a very old version. They are used to indicate, whether a PDF exists. Before we fix it, you might circumvent this bug by setting `file_hash` to some non-null value:

```SQL
UPDATE items SET file_hash = 'placeholder'
```

**Comment by oraclesod** [comment]
Ok i have applied that to the DB, now the PDF and download button can be clicked on each library item, and the download works, but the built in PDF viewer still fails with a 500 (internal server error)

**Comment by Aktarus006** [comment]
Hello,

I had to work on the migration from a 2.1 version of I,Librarian to the actual 5.5 at my work.

What I did:
- First I downloaded 4.10 and migrate the database from 2.1 to 4.10
- Then I installed the 5.5 and migrate the 4.10 database. I also moved the pdfs to /var/www/librarian/data/pdfs

At this point, I had the same issue as you. So I filled the file_hash with a random value for all my items.
The PDF button was now enabled but the PDF was not displayed.

My old PDFS were named with only five digits in comparison of the nine of the current version so I renamed them adding '0000' at the start of each file.
I also created folders to split these files as it does in this version. Create a folder named by the first 3 digits on the files (000) and inside it folders with the 3 next digits. And  I moved the pdfs accordingly.

Now, It is almost working properly. Just the "manage pdf" link returns an error and the pdf viewer from I,librarian is not working I had to switch to the web browser one.

Any clue on how to correct it?

Thanks for the help!

**Comment by mkucej** [comment]
I recommend setting the `error_messages` `level` to `debug` in ilibrarian.ini and get more detailed information about your errors. Please open a new issue for each distinct error. Thank you.

**Comment by oraclesod** [comment]
mkucej, should i follow Aktarus006's steps to get to the same place ? I have thousands of files so not sure how many to put in each folder and how to number them....

**Comment by mkucej** [comment]
You should be able to fix hashes in 5.6.0 like this:

1. Null file hashes:
```SQL
UPDATE items SET file_hash = NULL
```
2. Run `Administrator` > `Find duplicates` > `Identical PDFs`. Correct file hashes will be created.

**Comment by oraclesod** [comment]
Thank you so much, I have it all working now.
